### PR TITLE
Refactor- Enum class for payment intent status constants

### LINKED ISSUE #5327: Create and use Payment Intent Status Enum class for payment Intent Status constants
### Description

Stripe has distinct [payment intent statuses](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-status). Across our codebase, we are currently checking for payment intents as strings. (examples - [one](https://github.com/Automattic/woocommerce-payments/blob/59715410958dbcd4588ad2a16e6b46012685447f/tests/unit/test-class-wc-payments-order-service.php#L426), [two](https://github.com/Automattic/woocommerce-payments/blob/59715410958dbcd4588ad2a16e6b46012685447f/tests/unit/test-class-wc-payments-order-service.php#L297), [three](https://github.com/Automattic/woocommerce-payments/blob/59715410958dbcd4588ad2a16e6b46012685447f/tests/unit/test-class-wc-payments-order-service.php#L317) ) 

### Acceptance criteria

Payment Intent statuses should be moved into their own Enum class and the codebase should use constants, instead of hard-coded strings. 

This would be similar to what we are doing for [Order Statuses here.](https://github.com/Automattic/woocommerce-payments/pull/5320) 

### Testing instructions
The changes should clear existing unit tests. 

### Additional context
Inspired by 
https://github.com/Automattic/woocommerce-payments/issues/5286

### PR DESCRIPTION
**Pull request by nagpai**
Fixes #5327 

#### Changes proposed in this Pull Request

The PR refactors all typed strings of Payment Intent status to use constants, saved within an enum class - `WCPay\Constants\Payment_Intent_Status` . 

The change is similar to what we did for [Order Statuses recently](https://github.com/Automattic/woocommerce-payments/pull/5320). 

#### Testing instructions

* Existing unit tests should run without errors. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply) - Not applicable

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Review by elazzabi — APPROVED** [review]
Looks good! Thank you for working on this!

File: includes/constants/class-payment-intent-status.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * This class gives a list of all Payment Intent status name constants.
+ * ref: https://stripe.com/docs/api/payment_intents/object#payment_intent_object-status
+ *
+ * @psalm-immutable
+ */
+class Payment_Intent_Status extends Enum {
+	const REQUIRES_PAYMENT_METHOD = 'requires_payment_method';
+	const COMPLETED               = 'completed';
```
Comment (htdat) [inline includes/constants/class-payment-intent-status.php]: PaymentIntent does not have this `completed` status. Might double check again?

**Review by htdat — CHANGES_REQUESTED** [review]

File: includes/constants/class-payment-intent-status.php (hunk 1 above)
Comment (nagpai) [reply includes/constants/class-payment-intent-status.php]: Hi Dat  @htdat ! Thanks for your review. That was an eagle eye catch! Indeed, my oversight here, the status `completed` may have crept in from the `Order Status` class that I copied and amended here. 

I checked once again with the [Stripe docs ](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-status)on Payment Intent, and removed the unwanted status. 

I hope this looks fine now?
[2e4b1ab](https://github.com/Automattic/woocommerce-payments/pull/5345/commits/2e4b1abdfbc89ca57574d76bc671b32fa4108bf9)

File: includes/class-wc-payment-gateway-wcpay.php [hunk 2]
```diff
@@ -63,7 +64,11 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 	 *
 	 * @type array
 	 */
-	const SUCCESSFUL_INTENT_STATUS = [ 'succeeded', 'requires_capture', 'processing' ];
+	const SUCCESSFUL_INTENT_STATUS = [
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:67]: I saw we use same set of statuses in `includes/admin/class-wc-rest-payments-orders-controller.php`, wondering if we should use the constant there as well, WDYT?

**Review by kalessil — CHANGES_REQUESTED** [review]
Looks good overall! See one final comment from me.

File: includes/class-wc-payment-gateway-wcpay.php (hunk 2 above)
Comment (nagpai) [reply includes/class-wc-payment-gateway-wcpay.php:67]: Thanks Vlad. In the PR, I have [replaced the status strings](https://github.com/Automattic/woocommerce-payments/pull/5345/files#diff-2ba6807096e19c71b79c31fee4cfd08e114f568917d337995bf0e9358b8056d5) with the constants in `includes/admin/class-wc-rest-payments-orders-controller.php`  . I hope I have not missed any lines there. I had done a string search across the codebase for each of the statuses, and made relevant replacements wherever it was a Payment Intent status.

Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:67]: I was referencing to

```php
				[
					Payment_Intent_Status::PROCESSING,
					Payment_Intent_Status::REQUIRES_CAPTURE,
					Payment_Intent_Status::SUCCEEDED,
				]
```

in `includes/admin/class-wc-rest-payments-orders-controller.php/` (2 occasions).

Those are the same array of statuses as in this constant, so we could use `WC_Payment_Gateway_WCPay::SUCCESSFUL_INTENT_STATUS` there.

Does it make sense?

Comment (nagpai) [reply includes/class-wc-payment-gateway-wcpay.php:67]: Makes complete sense. Thanks for explaining this in detail! 

I have made the changes here 
51f0f75

Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:67]: @nagpai, this looks better, can we keep a one-line formatting style for the calls there (as it was initially)?

Comment (nagpai) [reply includes/class-wc-payment-gateway-wcpay.php:67]: I have changed it [here](https://github.com/Automattic/woocommerce-payments/pull/5345/files#diff-2ba6807096e19c71b79c31fee4cfd08e114f568917d337995bf0e9358b8056d5R151) and [here](https://github.com/Automattic/woocommerce-payments/pull/5345/files#diff-2ba6807096e19c71b79c31fee4cfd08e114f568917d337995bf0e9358b8056d5R254). 

 Thank you!

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 3]
```diff
@@ -758,7 +759,7 @@ public function is_setup_intent_success_creation_redirection() {
 		return ! empty( $_GET['setup_intent_client_secret'] ) && // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 			! empty( $_GET['setup_intent'] ) && // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 			! empty( $_GET['redirect_status'] ) && // phpcs:ignore WordPress.Security.NonceVerification.Recommended
-			'succeeded' === $_GET['redirect_status']; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			Payment_Intent_Status::SUCCEEDED === $_GET['redirect_status']; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Comment (htdat) [inline includes/payment-methods/class-upe-payment-gateway.php]: `redirect_status` may not be related to Payment_Intent_Status?

File: includes/constants/class-payment-intent-status.php (hunk 1 above)
Comment (htdat) [reply includes/constants/class-payment-intent-status.php]: @nagpai - Looks good now! Thanks for the review as well as this whole PR.

File: includes/payment-methods/class-upe-payment-gateway.php (hunk 3 above)
Comment (kalessil) [reply includes/payment-methods/class-upe-payment-gateway.php]: :+1:

**Review by kalessil — APPROVED** [review]
LGTM!

**Comment by kalessil** [comment]
@htdat: we are ready for a final review round.

**Comment by kalessil** [comment]
@nagpai after we merge this one, I suggest writing a post announcing the two enums introduction, so other teams are aware of the enums existence. WDYT?

**Review by htdat — APPROVED** [review]

**Comment by nagpai** [comment]
@htdat , @kalessil , @elazzabi  - Thank you so much for your reviews! 

>@nagpai after we merge this one, I suggest writing a post announcing the two enums introduction, so other teams are aware of the enums existence. WDYT?

This is a great idea! I will write a post! Thank you!

**Pull request merged**